### DISCUSSION TITLE
Overlapping community detection?

### DISCUSSION THREAD

**Discussion by jasperhyp** [networkx/networkx#5453] [Algorithms]
Are there any functions that could allow for overlapping community detection in `networkx`?

**Comment by dschult** [comment]
You will need to explain your terms more for us to be able to help you effectively.
Our community detection algorithms do not produce overlapping communities. But you can obviously create communities based on different edge weight attributes and these communities are likely to be overlapping.

**Reply by jasperhyp** [reply]
Thanks for the response.  By "overlapping community detection", I am referring to _softly_ "clustering" the vertices based on the network structure and **1 single node attribute** in a way that generates communities of vertices where 1 single vertice could appear in more than one community.  A visualization of the difference between the expected results of regular community detection and overlapping community detection, respectively, is shown below (source: https://www.sciencedirect.com/science/article/abs/pii/S1084804518300560).

![image](https://user-images.githubusercontent.com/52191793/161385353-a330006f-d034-44a7-b7b8-56881fc3f478.png)

**Reply by dschult** [reply]
Cool...  Do you know of community detection algorithms that allow for this?

**Reply by jasperhyp** [reply]
I just ran into a problem that requires this kind of overlapping CD, and I found two surveys that probably cover most of the SOTA (speaking of SOTA, it might be really risky to call a clustering algorithm SOTA) algos by the time they were written: https://www.cs.rpi.edu/~szymansk/papers/acm-cs.13.pdf, https://appliednetsci.springeropen.com/articles/10.1007/s41109-020-00289-9.  Happy to implement one or two in Python (e.g. the probably simplest one, overlapping version of Girvan-Newman) but I might not be very familiar with implementing cpp in Python.

Summary of algos from the first survey:
![image](https://user-images.githubusercontent.com/52191793/161400913-6ff65b4e-7e14-4c84-99e4-f3e8243a6436.png)

**Reply by dschult** [reply]
The first article suggests that some of these methods are extensions of the partition community detection algorithms -- e.g. Newman and Girvan method is for the partition problem, but it has been extended by two papers: Nicosia et al (2009) and Shen (2103) so that it can work for the overlapping methods.

My current understanding of  the overlapping CD problem is that each node has a membership total of 1 that it can distribute to the communities. So they can be 20% in community 1, 30% in 2 and 50% in 3.  The goal is to pick those values to maximize some definition of modularity.  

NetworkX does not have any algorithms that solve this problem, but you might be able to start from our implementation of a partition algorithm and extend it to the overlapping case in the way that e.g. EAGLE did -- and not have to reconstruct the entire algorithm in Python -- only the changed parts.

**Reply by jasperhyp** [reply]
Yes, I agree that many of the algos for this problem are adapted from some typical partition methods, and adding the notion of "membership strength".  Thanks for the advice,  I looked into `networkx`'s source code of Girvan-Newman and Louvain and they seem to be pretty straightforward.  Will try to adapt from the current implementations.

**Reply by kpetridis24** [reply]
I don't know if this helps but, in one of my previous projects, for simplicity and testing purposes i tried the following method. At first i located a sub-community per node of the graph using random walkers. Then, i started a joining process, where i joined pairs of sub-communities (clusters) to form bigger clusters based on similarity. This means that, if the similarity between two communities exceeded a certain threshold, they are joined into one. The process is repeated until no pair is found so that they are more similar than what is allowed. Of cource it's not supposed to be the best method by any means, but it helped me run some tests and visualize the effect!
